src/utils.rs

```diff
@@ -154,34 +22,50 @@ where
 
 /// Returns error if index is not at a valid utf-8 char boundary
 #[inline]
-pub fn is_char_boundary<const N: usize>(s: &ArrayString<N>, idx: u8) -> Result<(), Utf8>
+pub fn is_char_boundary<const N: usize>(
+    s: &ArrayString<N>,
+    idx: impl Into<usize>,
+) -> Result<(), Utf8>
 where
     ArrayString<N>: ValidCapacity,
 {
+    let idx = idx.into();
     trace!("Is char boundary: {} at {}", s.as_str(), idx);
-    if s.as_str().is_char_boundary(idx.into()) {
+    if s.as_str().is_char_boundary(idx) {
         return Ok(());
     }
     Err(Utf8)
 }
 
+#[inline]
+pub unsafe fn is_char_boundary_at(arr: &[u8], index: usize) -> bool {
+    if index == 0 {
+        return true;
+    }
+    (*arr.get_unchecked(index) as i8) >= -0x40
```

Comment:
Is this really better than using stdlib's implementation for &str, I think I'm missing the reason for arr to be of `&[u8]`

But at least could we copy https://doc.rust-lang.org/src/core/str/mod.rs.html#211?

Reply:
we can make it 100% safe, but my testing showed it wasn't always able to optimize out the range check.
i'm open to making it 100% safe, but we need better benchmarks to compare. maybe i'll test that on the other branch.
There is also the #[no_panic] proc_macro crate to compile time check if the function can panic (all panics have indeed been optimized out)

Reply:
Depending on the performance difference I can live with the loss for removing unsafe. Is this really more efficient than stdlibs'? Anyway, at least using the byte function to see if it's utf8 instead of the hardcoded integer. But again, I can change that later. I will still need time to test this and review it properly.

The no_panic macro didn't really work when I tried, maybe time has done good to it.

Reply:
well no_panic caught a few interesting ones. the catch is that it works only in release mode with O3, so i enable it only if debug_assertions are disabled. so right now only the benchmarks can see the panics. we could add an additional feature gate too.